compose-builder/tui-generator.sh

```diff
@@ -0,0 +1,299 @@
+#  *
+#  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+#  * in compliance with the License. You may obtain a copy of the License at
+#  *
+#  * http://www.apache.org/licenses/LICENSE-2.0
+#  *
+#  * Unless required by applicable law or agreed to in writing, software distributed under the License
+#  * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  * or implied. See the License for the specific language governing permissions and limitations under
+#  * the License.
+#  *
+#  *******************************************************************************/
+
+# generator.sh - provide a Terminal User Interface to make generation of different EdgeX Foundry
+# services more comprehensible.
+# Author: Shantanoo (Shan) Desai <shantanoo.desai@gmail.com>
+
+# Map the available options in `make compose` either as standalone strings or array of strings
+
+# Individual Compose Descriptions that may not fall into EdgeX-Foundry Architecture Components
+# Or may be an initial "out-of-the-box" configuration
+
+# Check if whiptail exists to render the UI, if not, exit script
+WHIPTAIL=$(which whiptail)
+
+if [ -z $WHIPTAIL ]
+then
+    echo "This script requires whiptail to render the UI."
+    exit 1;
+fi
+
+MAKE=/usr/bin/make
+
+SELECTED_DEVSERVICES=()
+SELECTED_APPSERVICES=()
+SELECTED_BUS=()
+SELECTED_OTHERS=()
+
+## Additional Options
+additionalOpts=(
+    "mqtt-broker"
+    "modbus-sim"
+)
+
+## General Options Description
+declare -A additionalOptsDesc=(
+    [modbus-sim]="Include ModBus Simulator"
+    [mqtt-broker]="Include MQTT Broker"
+)
+
+
+## Available App Services in EdgeX-Foundry
+appServices=(
+    "asc-http"
+    "asc-mqtt"
+    "asc-sample"
+    "as-llrp"
+    "asc-ex-mqtt"
+)
+
+## App Service Descriptions
+declare -A appServiceDesc=(
+    [asc-http]="Include HTTP Export App Service"
+    [asc-mqtt]="Include MQTT Export App Service"
+    [asc-sample]="Include Sample App Service"
+    [as-llrp]="Include RFID LLRP Inventory"
+    [asc-ex-mqtt]="Include External MQTT Trigger App Service"
+)
+
+
+## Available Device Services in EdgeX-Foundry
+deviceServices=(
+    "ds-bacnet"
+    "ds-camera"
+    "ds-grove"
+    "ds-modbus"
+    "ds-mqtt"
+    "ds-rest"
+    "ds-snmp"
+    "ds-virtual"
+    "ds-coap"
+    "ds-gpio"
+    "ds-llrp"
+)
+
+## Device Service Descriptions
+declare -A deviceServiceDesc=(
+    [ds-bacnet]="Include BACnet Device Service"
+    [ds-camera]="Include Camera Device Service"
+    [ds-grove]="Include Grove Device Service (valid only: arm64)"
+    [ds-modbus]="Include ModBus Device Service"
+    [ds-mqtt]="Include MQTT Device Service"
+    [ds-rest]="Include REST Device Service"
+    [ds-snmp]="Include SNMP Device Service"
+    [ds-virtual]="Include Virtual Device Service"
+    [ds-coap]="Include CoAP Device Service"
+    [ds-gpio]="Include GPIO Device Service"
+    [ds-llrp]="Include RFID LLRP Device Service"
+
+)
```

Comment:
You can use a nifty bash trick to reduce the amount of duplicated code.

Use `for index in "${!deviceServiceDesc[@]}";` to loop through the keys in the `deviceServiceDesc` associative array. This will allow you to remove the `deviceServices` array altogether.

Rinse and repeat for all similar code blocks.